doc/designs/quic-design/dgram-api.md

```diff
@@ -0,0 +1,322 @@
+Datagram BIO API revisions for sendmmsg/recvmmsg
+================================================
+
+We need to evolve the API surface of BIO which is relevant to BIO_dgram (and the
+eventual BIO_dgram_mem) to support APIs which allow multiple datagrams to be
+sent or received simultaneously, such as sendmmsg(2)/recvmmsg(2).
+
+Options for the API surface include:
+
+sendmmsg/recvmmsg-like API
+--------------------------
+
+```c
+int BIO_readm(BIO *b, BIO_mmsghdr *msgvec,
+              unsigned len, int flags, struct timespec *timeout);
+int BIO_writem(BIO *b, BIO_mmsghdr *msgvec,
+              unsigned len, int flags, struct timespec *timeout);
+```
+
+We can either define `BIO_mmsghdr` as a typedef of `struct mmsghdr` or redefine
+an equivalent structure. The former has the advantage that we can just pass the
+structures through to the syscall without copying them.
+
+Note that in `BIO_mem_dgram` we will have to process and therefore understand
+the contents of `struct mmsghdr` ourselves. Therefore, initially we define a
+subset of `struct mmsghdr` as being supported, specifically no control messages;
+`msg_name` and `msg_iov` only.
+
+The flags argument is defined by us. Initially we can support something like
+`MSG_DONTWAIT` (say, `BIO_DONTWAIT`).
+
+### Implementation Questions
+
+If we go with this, there are some issues that arise:
+
+- Are `BIO_mmsghdr`, `BIO_msghdr` and `BIO_iovec` simple typedefs
+  for OS-provided structures, or our own independent structure
+  definitions?
```

Comment:
I think the lessons we can draw from 32-bit `time_t`, `timegm()` non-portability, etc. indicate that we should "own" the semantics of our public types, and having the public type defined as a typedef (even within an `#ifdef`) can cause problems in the future.  But we should feel free to have our structure layout match the OS one such that a compiler could optimize the "copy the contents" operation away.

Reply:
Yes, agreed; note that these questions have already been largely answered. See #18270, which includes manpages, as well as #18238.